File: webui/src/test/scala/com/taskadapter/config/JsonFactoryTest.scala
package com.taskadapter.config

import com.taskadapter.connector.common.FieldMappingBuilder.getMapping
import com.taskadapter.model.{CustomDate, CustomFloat, CustomString, DoneRatio, Summary}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

@RunWith(classOf[JUnitRunner])
class JsonFactoryTest extends FunSpec with Matchers {

  it("can save and load fields") {
    val fieldMappings = Seq(
      getMapping(Summary),
      getMapping(DoneRatio),
      getMapping(CustomDate("f")),
      getMapping(CustomFloat("f")),
      getMapping(CustomString("f"))
    )
    val str = JsonFactory.toString(fieldMappings.toSeq)
    val parsed = JsonFactory.fromJsonString(str)
    parsed shouldBe fieldMappings
  }
}

File: webui/src/test/java/com/taskadapter/config/JsonFactoryTest.java
package com.taskadapter.config;

import com.taskadapter.common.ui.FieldMapping;
import com.taskadapter.common.ui.NewConfigSuggester;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.CustomDate;
import com.taskadapter.model.CustomFloat;
import com.taskadapter.model.CustomString;
import org.junit.Test;

import java.util.List;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

public class JsonFactoryTest {
    @Test
    public void canSaveAndLoadFields() {
        List<FieldMapping<?>> fieldMappings = List.of(
                NewConfigSuggester.duplicateFieldIntoMapping(AllFields.summary),
                NewConfigSuggester.duplicateFieldIntoMapping(AllFields.doneRatio),
                NewConfigSuggester.duplicateFieldIntoMapping(new CustomDate("f")),
                NewConfigSuggester.duplicateFieldIntoMapping(new CustomFloat("f")),
                NewConfigSuggester.duplicateFieldIntoMapping(new CustomString("f"))
        );
        var str = JsonFactory.toString(fieldMappings);
        var parsed = JsonFactory.fromJsonString(str);
        assertThat(parsed).isEqualTo(fieldMappings);
    }
}
